circuits/cpp/src/aztec3/circuits/kernel/private/common.cpp

```diff
@@ -44,6 +49,58 @@ void common_validate_call_stack(DummyComposer& composer, PrivateCallData<NT> con
     }
 }
 
+
+/**
+ * @brief Validate all read requests against the historic private data root.
+ * Use their membership witnesses to do so. If the historic root is not yet
+ * initialized, initialize it using the first read request here (if present).
+ *
+ * @details More info here:
+ * - https://discourse.aztec.network/t/to-read-or-not-to-read/178
+ * - https://discourse.aztec.network/t/spending-notes-which-havent-yet-been-inserted/180
+ *
+ * @param composer
+ * @param read_requests the commitments being read by this private call
+ * @param read_request_membership_witnesses used to compute the private data root
+ * for a given request which is essentially a membership check.
+ * @param historic_private_data_tree_root This is a reference to the historic root which is
+ * an output (public input) of the kernel. All read requests much match this root. The very first
+ * read request in the the tx will initialize this root (in this function).
+ */
+void common_validate_read_requests(DummyComposer& composer,
+                                   NT::fr const& storage_contract_address,
+                                   std::array<fr, READ_REQUESTS_LENGTH> const& read_requests,
+                                   std::array<MembershipWitness<NT, PRIVATE_DATA_TREE_HEIGHT>,
+                                              READ_REQUESTS_LENGTH> const& read_request_membership_witnesses,
+                                   NT::fr const& historic_private_data_tree_root)
+{
+    auto num_read_requests = array_length(read_requests);
+    if (num_read_requests > 0) {
```

Comment:
Membership checks must be performed for all `READ_REQUESTS_LENGTH` reads even if `read_requests` is partially filled, but the result of checks for zeros in `read_requests` must be "ignored". Something to keep in mind for the real circuit version of things.